File: MendeleyKit/MendeleyKit/Public Interface/API Interfaces/MendeleyRecommendationsAPI.m
/*
 ******************************************************************************
 * Copyright (C) 2014-2017 Elsevier/Mendeley.
 *
 * This file is part of the Mendeley iOS SDK.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *****************************************************************************
 */

#import "MendeleyRecommendationsAPI.h"
#import "MendeleyRecommendedArticle.h"
#import "NSError+Exceptions.h"

@implementation MendeleyRecommendationsAPI

#pragma mark Private methods

- (NSDictionary *)defaultServiceRequestHeaders
{
    return @{ kMendeleyRESTRequestContentType: kMendeleyRESTRequestJSONRecommendationsType };
}

- (NSDictionary *)feedbackServiceRequestHeaders
{
    return @{ kMendeleyRESTRequestContentType: kMendeleyRESTRequestJSONRecommendationFeedbackType };
}

- (NSDictionary *)feedbackBodyParametersWithTrace:(NSString *)trace
                                         position:(NSNumber *)position
                                       userAction:(NSString *)userAction
                                         carousel:(NSNumber *)carousel
{
    return @{ kMendeleyJSONTrace: trace,
              kMendeleyJSONPosition: position,
              kMendeleyJSONUserAction: userAction,
              kMendeleyJSONCarousel: carousel };
}

#pragma mark -

- (MendeleyTask *)recommendationsBasedOnLibraryArticlesWithParameters:(MendeleyRecommendationsParameters *)queryParameters
                                                                 task:(MendeleyTask *)task
                                                      completionBlock:(MendeleyArrayCompletionBlock)completionBlock
{
    [NSError assertArgumentNotNil:completionBlock argumentName:@"completionBlock"];
    
    [self.provider invokeGET:self.baseURL
                         api:kMendeleyRESTAPIRecommendationsBasedOnLibrary
           additionalHeaders:[self defaultServiceRequestHeaders]
             queryParameters:[queryParameters valueStringDictionary]
      authenticationRequired:YES
                        task:task
             completionBlock:^(MendeleyResponse * _Nullable response, NSError * _Nullable error) {
                 MendeleyBlockExecutor *blockExec = [[MendeleyBlockExecutor alloc] initWithArrayCompletionBlock:completionBlock];
                 if (![self.helper isSuccessForResponse:response error:&error])
                 {
                     [blockExec executeWithArray:nil
                                        syncInfo:nil
                                           error:error];
                 }
                 else
                 {
                     MendeleyModeller *jsonModeller = [MendeleyModeller sharedInstance];
                     if (response.responseBody[kMendeleyJSONData] != nil)
                     {
                         [jsonModeller parseJSONData:response.responseBody[kMendeleyJSONData] expectedType:kMendeleyModelRecommendations completionBlock: ^(NSArray<MendeleyRecommendedArticle *> *recommendationsList, NSError *parseError) {
                             if (nil != parseError)
                             {
                                 [blockExec executeWithArray:nil
                                                    syncInfo:nil
                                                       error:parseError];
                             }
                             else
                             {
                                 [blockExec executeWithArray:recommendationsList
                                                    syncInfo:response.syncHeader
                                                       error:nil];
                             }
                         }];
                     }
                 }

             }];
    
    return nil;
}

- (MendeleyTask *)feedbackOnRecommendation:(NSString *)trace
                                  position:(NSNumber *)position
                                userAction:(NSString *)userAction
                                  carousel:(NSNumber *)carousel
                                      task:(MendeleyTask *)task
                           completionBlock:(MendeleyCompletionBlock)completionBlock
{
    [NSError assertArgumentNotNil:trace argumentName:@"trace"];
    [NSError assertArgumentNotNil:position argumentName:@"position"];
    [NSError assertArgumentNotNil:userAction argumentName:@"userAction"];
    [NSError assertArgumentNotNil:carousel argumentName:@"carousel"];
    [NSError assertArgumentNotNil:completionBlock argumentName:@"completionBlock"];
    
    [self.provider invokePOST:self.baseURL
                          api:kMendeleyRESTAPIRecommendationFeedback
            additionalHeaders:[self feedbackServiceRequestHeaders]
               bodyParameters:[self feedbackBodyParametersWithTrace:trace position:position userAction:userAction carousel:carousel]
                       isJSON:YES
       authenticationRequired:YES
                         task:task
              completionBlock:^(MendeleyResponse * _Nullable response, NSError * _Nullable error) {
                  MendeleyBlockExecutor *blockExec = [[MendeleyBlockExecutor alloc] initWithCompletionBlock:completionBlock];
                  if (![self.helper isSuccessForResponse:response error:&error])
                  {
                      [blockExec executeWithBool:NO
                                           error:error];
                  }
                  else
                  {
                      [blockExec executeWithBool:YES
                                           error:nil];
                  }
              }];
    
    return nil;
}

@end

File: MendeleyKit/MendeleyKit/Public Interface/API Interfaces/MendeleyRecommendationsAPI.swift
//
//  MendeleyRecommendationsAPI.swift
//  MendeleyKit
//
//  Created by Tunney, Aaron (ELS) on 20/02/2018.
//  Copyright © 2018 Mendeley. All rights reserved.
//

import UIKit

@objc public class MendeleyRecommendationsAPI: MendeleyObjectAPI {
    @objc public func recommendationsBasedOnLibraryArticles(withParameters parameters: MendeleyRecommendationsParameters?, task: MendeleyTask?, completionBlock:  MendeleyArrayCompletionBlock?) {
      
        provider.invokeGET(baseURL,
                           api: kMendeleyRESTAPIRecommendationsBasedOnLibrary,
                           additionalHeaders: defaultServiceRequestHeaders,
                           queryParameters: parameters?.valueStringDictionary(),
                           authenticationRequired: true,
                           task: task) { (response, error) in
                            let blockExec = MendeleyBlockExecutor(arrayCompletionBlock: completionBlock)
                            
                            if (try? self.helper.isSuccess(for: response)) == nil {
                                blockExec?.execute(with: nil, syncInfo: nil, error: error)
                            } else {
                                if let responseBody = response?.responseBody as? [String: Any] {
                                    if let json = responseBody[kMendeleyJSONData] {
                                        do {
                                            let data = try JSONSerialization.data(withJSONObject: json, options: [])
                                            let decoder = JSONDecoder()
                                            let recommendedArticles = try decoder.decode([MendeleyRecommendedArticle].self, from: data)
                                            blockExec?.execute(with: recommendedArticles, syncInfo: response?.syncHeader, error: nil)
                                        } catch {
                                            print(error)
                                            blockExec?.execute(with: nil, syncInfo: nil, error: error)
                                        }
                                    }
                                }
                            }
        }
    }
    
    @objc public func feedbackOnRecommendation(_ trace: String, position: Int, userAction: String, carousel: Int, task: MendeleyTask?, completionBlock: MendeleyCompletionBlock?) {
        
        let bodyParameters = feedbackBodyParamaters(withTrace: trace, position: position, userAction: userAction, carousel: carousel)
        
        provider.invokePOST( baseURL,
                            api: kMendeleyRESTAPIRecommendationFeedback,
                            additionalHeaders: feedbackServiceRequestHeaders,
                            bodyParameters: bodyParameters,
                            isJSON: true,
                            authenticationRequired: true,
                            task: task) { (response, error) in
                                let blockExec = MendeleyBlockExecutor(completionBlock: completionBlock)

                                if (try? self.helper.isSuccess(for: response)) == nil {
                                    blockExec?.execute(with: false, error: error)
                                } else {
                                    blockExec?.execute(with: true, error: nil)
                                }
        }
    }
    
    // Mark: Headers and Query Parameters
    
    private let defaultServiceRequestHeaders = [kMendeleyRESTRequestContentType: kMendeleyRESTRequestJSONRecommendationsType]
    private let feedbackServiceRequestHeaders = [kMendeleyRESTRequestContentType: kMendeleyRESTRequestJSONRecommendationFeedbackType]
    
    func feedbackBodyParamaters(withTrace trace: String, position: Int, userAction: String, carousel: Int) -> [String: Any] {
        return [kMendeleyJSONTrace: trace,
                kMendeleyJSONPosition: position,
                kMendeleyJSONUserAction: userAction,
                kMendeleyJSONCarousel: carousel]
    }
}
